tty: add ref() so process.stdin.ref() etc. work

insightfuls opened the pull request:
##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] `make -j4 test` (UNIX) or `vcbuild test nosign` (Windows) passes
- [X] a test and/or benchmark is included
- [X] the commit message follows commit guidelines
##### Affected core subsystem(s)

<!-- provide affected core subsystem(s) (like doc, cluster, crypto, etc) -->

tty
##### Description of change

<!-- provide a description of the change below this comment -->

tty.ReadStream and tty.WriteStream are documented to be net.Socket subclasses, so should support unref() and ref(), but only unref() worked. I ran into this problem when following advice at http://stackoverflow.com/questions/26004519/why-doesnt-my-node-js-process-terminate-once-all-listeners-have-been-removed to get Node to exit when I stop listening to stdin (which seems to be delayed until a newline is received otherwise). If I changed my mind and wanted to start listening again prior to exiting, it wasn't possible as ref() wasn't implemented (net.Socket.ref threw when trying to call tty_wrap.ref which didn't exist). This fixes it up trivially by implementing tty_wrap.ref().

Fishrock123 commented:
IIRC when I last tried this it didn't actually undo the weak reference..

Fishrock123 commented:
CI... https://ci.nodejs.org/job/node-test-pull-request/3066/

insightfuls commented:
It works for me with this test script:

```
"use strict";
process.stdout.write("Keep Node open [y/N]?: ");
process.stdin.on("readable", function() {
	var data = process.stdin.read();
	if (data === null) {
		return;
	}
	if (data[0] === "Y".charCodeAt(0) || data[0] === "y".charCodeAt(0)) {
		process.stdin.ref();
	} else {
		process.stdin.unref();
		process.stdout.write("There will be a 5-second delay for you to change your mind.\n");
		setTimeout(function(){}, 5000);
	}
	process.stdout.write("Keep Node open [y/N]?: ");
});
process.on("exit", function(){ process.stdout.write("\n") });
```

AFAICT, I can't work something like this into the automated tests, but if there's a way to do so, I can try; just point me in the right direction.

Re the CI, is there something that requires my action? I realise it's not passing, but I'm not sure what all the failures are. I do know one test was failing before I made my changes, and I rebased before submitting the PR, somewhat blindly following the instructions in the Contributing document, as I'm somewhat a Git n00b. Should I rebase my branch onto a commit where all the tests are passing? Or is there something else I should do? There seem to also be linting problems showing up in Jenkins which I doubt have anything to do with this change, but I'm not sure how to get more detailed information to even find out. I'm happy to do whatever I can, but I'll need a little guidance to know what to do.

Fishrock123 commented:
Ok, there is a way to make the test work and I did verify that it works correctly.

Put this is `test/pseudo-tty/` -- you'll also need a file of the same name with the extension `.out`, it can just have a single newline in it. that is ok.

```js
'use strict';
require('../common');

const TTY = process.binding('tty_wrap').TTY;

const handle = new TTY(0);
handle.readStart();

handle.unref();
handle.ref();
handle.unref();
```

insightfuls commented:
Thanks. That was the start I needed. Looking at the code you provided, though, it only truly tests unref(): ref() could be a no-op and the test would still pass (unless it did something like throw, as it used to, but testing it doesn't throw is covered by the test I already provided). I've reworked it to use an unref()'d timeout to check that ref() does indeed keep Node running, so works as expected. If it doesn't, the timeout won't keep Node open either so there will be no output. Of course this also tests unref() works, because otherwise Node would hang and not exit at all.

`make -j4 test` is working for me; though one test in the suite seems to fail non-deterministically (ouch), it is nothing to do with this change (something to do with watching the filesystem).

Is there anything more I need to do here?

Fishrock123 commented on test/parallel/test-tty-ref.js:
```diff
@@ -0,0 +1,15 @@
+var handle = new TTY(1);
+
+handle.unref();
+handle.ref(); // Should not throw.
```
Please remove this test - it will always be skipped and is therefore not useful. :)

Fishrock123 replied:
(`pseudo-tty/` tests are runs as party of normal test runs)

cjihrig commented on test/pseudo-tty/ref_and_unref.js:
```diff
@@ -0,0 +1,16 @@
+handle.unref();
+handle.ref();
+
+var timeout = setTimeout(function() {
```
Can we not use a timer. They tend to make tests flakey. Is there anything in `process._getActiveHandles()` that can be checked here instead?

Fishrock123 replied:
If not, `const` and an arrow function here would be appriciated

Fishrock123 replied:
@cjihrig Hmmm, I like the timer check here tbh.

I think `_getActiveHandles()` will be hard to filter though since there will also be process.stdout at some point, and maybe a timer handle if we don't use `_unrefActive()`

(I've never had much checking `_getActiveHandles()` in a useful way I guess?)

cjihrig replied:
Wouldn't something like this work?

```javascript
function isHandleActive(handle) {
  return process._getActiveHandles().some((active) => {
    return active === handle;
  });
}
```

Fishrock123 commented on test/pseudo-tty/ref_and_unref.js:
```diff
@@ -0,0 +1,16 @@
+  console.log('still running');
+  handle.unref();
+}, 10);
+timeout.unref();
```
Maybe this is nitpicking but I'd feel better about testing this with an internal timer:

```js
const timers = require('timers');

// ...

const timer = {
  _onTimeout() {
    console.log('still running');
    handle.unref();
  }
};
timers.enroll(timer, 50);
timers._unrefActive(timer)
```

Fishrock123 commented:
Looking good, CI: https://ci.nodejs.org/job/node-test-pull-request/3242/

Fishrock123 commented:
@insightfuls Looks like lots of errors!

```
not ok 1162 pseudo-tty/ref_and_unref
# /Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node[62514]: ../src/async-wrap-inl.h:109:v8::Local<v8::Value> node::AsyncWrap::MakeCallback(const v8::Local<v8::String>, int, v8::Local<v8::Value> *): Assertion `cb_v->IsFunction()' failed.
# 1: node::Abort() [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 2: node::RunMicrotasks(v8::FunctionCallbackInfo<v8::Value> const&) [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 3: node::StreamBase::EmitData(long, v8::Local<v8::Object>, v8::Local<v8::Object>) [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 4: node::StreamWrap::OnReadImpl(long, uv_buf_t const*, uv_handle_type, void*) [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 5: node::StreamWrap::OnReadCommon(uv_stream_s*, long, uv_buf_t const*, uv_handle_type) [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 6: uv__stream_eof [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 7: uv__stream_io [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 8: uv__io_poll [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 9: uv_run [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 10: node::Start(int, char**) [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
# 11: start [/Users/iojs/build/workspace/node-test-commit-osx/nodes/osx1010/out/Release/node]
```

@cjihrig Any idea? I'm not really sure why `MakeCallback` would fail from this.

@insightfuls Maybe it would better to use Colin's suggestion for testing this.

cjihrig commented:
That's strange. Nothing really jumps out, and it passes locally on my mac.

Fishrock123 commented:
Same, it also passes locally for me.

addaleax commented:
I’d assume the reason it fails on CI is that standard input ends immediately there, and so the TTY handle tries to call `.onread()` to communicate that.

I am able to reproduce the failure using `echo -n | ./node_g test/pseudo-tty/ref_and_unref.js`, and adding a dummy `.onread()` should fix it:

```diff
diff --git a/test/pseudo-tty/ref_and_unref.js b/test/pseudo-tty/ref_and_unref.js
index 0faad3c..2131546 100644
--- a/test/pseudo-tty/ref_and_unref.js
+++ b/test/pseudo-tty/ref_and_unref.js
@@ -4,6 +4,7 @@ require('../common');
 const TTY = process.binding('tty_wrap').TTY;
 
 const handle = new TTY(0);
+handle.onread = () => {};
 handle.readStart();
 
 handle.unref();
```

insightfuls commented:
Ugh. `echo -n | make test` also fails. Turns out the Python test harness only puts a pty on stdout and stderr, not stdin, meaning in CI it wouldn't be testing a TTY at all. I've fixed that up.

I've updated my test case. I've opted for the internal timer approach, as I feel that tests precisely what I want it to test. I don't believe it will be flakey because it's not truly about time, it's only about giving the event loop an opportunity to tick over. But if any of you folk disagree, just see out the debate to get some consensus, and I'll be happy to follow it!

Additionally, though, the active handle aspect is, I believe, tested by 'test-handle-wrap-isrefed-tty', which I found in 'parallel', and wouldn't have been working, because it spawned a subprocess with piped stdio, so was really testing pipe-wrap not tty-wrap. I've moved it to pseudo-tty and removed the spawning so it truly tests TTYs now. I've followed the pattern from https://github.com/nodejs/node/pull/7613, doing the rename in a separate commit from fixing up the test.

Please see updated branch!

cjihrig commented:
I'm still -1 on using a timer that isn't really needed. If a ref'ed handle doesn't keep the event loop open, then Node is completely broken. Thoughts from @nodejs/testing?

Fishrock123 commented:
>  'test-handle-wrap-isrefed-tty', 

Crap. I forgot about that one when I fixed the others recently...

> Turns out the Python test harness only puts a pty on stdout and stderr, not stdin, meaning in CI it wouldn't be testing a TTY at all. I've fixed that up.

hmmm... I forgot about that. It should be possible to make it also get a fake stdin.

Fishrock123 commented:
CI Again: https://ci.nodejs.org/job/node-test-pull-request/3335/

I like this, but I'm fine switching to @cjihrig's way.

addaleax commented on test/pseudo-tty/test-handle-wrap-isrefed-tty.js:
```diff
@@ -0,0 +1,28 @@
+tty.unref();
+assert(tty._handle.hasRef(), false);
+tty.ref();
+assert(tty._handle.hasRef(), true);
```
I would prefer to use `assert.strictEqual` directly everywhere in this file; passing a fixed message for all `assert.strictEqual`s actually removes information from a possible error message.

insightfuls replied:
I feel like it might be better to leave it as is for consistency with parallel/test-handle-wrap-isrefed.js. This test case is basically one part of that file that needed to be separated out as it needs a TTY.

jasnell replied:
I agree with @addaleax ... we've been making an effort to convert as many tests as possible to using the strict assertion form where possible. As long as we're in this file making changes, we should go ahead and make the switch here. We can fix up the other tests for consistency separately

addaleax commented on test/pseudo-tty/ref_keeps_node_running.js:
```diff
@@ -0,0 +1,26 @@
+}
+
+const handle = new TTY(0);
+handle.readStart();
```
Please add a `handle.onread`, even if it works on CI right now without it.

addaleax commented on test/pseudo-tty/ref_keeps_node_running.js:
```diff
@@ -0,0 +1,26 @@
+const timers = require('timers');
+
+if (isTTY(0) == false) {
+  console.log('fd 0 is not a TTY.');
```
Using `common.skip` instead of `console.log` seems to make sense here.

insightfuls replied:
Yeah, the assertion is actually what I did after reading your original comment. 😄

addaleax commented on test/pseudo-tty/ref_keeps_node_running.js:
```diff
@@ -0,0 +1,26 @@
+const isTTY = process.binding('tty_wrap').isTTY;
+const timers = require('timers');
+
+if (isTTY(0) == false) {
```
nit: `===`

addaleax commented:
Looking good in general, and the last CI was green.

insightfuls commented:
Sorry about the double-update. I forgot to lint the first time. I believe I've addressed all the review comments now, except for changing the code which I really just moved from parallel to pseudo-tty and I'm not convinced I should be changing (or if I should, I should probably also change parallel/test-handle-wrap-isrefed.js, and possibly in a separate pull request).

Fishrock123 commented on test/pseudo-tty/ref_keeps_node_running.js:
```diff
@@ -0,0 +1,28 @@
+const isTTY = process.binding('tty_wrap').isTTY;
+const strictEqual = require('assert').strictEqual;
+
+strictEqual(isTTY(0), true, 'fd 0 is not a TTY');
```
`new TTY(0)` will abort if it isn't a TTY, this may not be necessary.

insightfuls replied:
It doesn't for me. I replaced the `strictEqual(isTTY(0), ...)` line with `console.log(isTTY(0))` and I get `false` output but no failure output and a success exit code when I pipe something in (such as `echo -n`). So I think this assertion is valuable. 😉

Fishrock123 replied:
huh... I wonder when that changed...

Fishrock123 commented:
CI again: https://ci.nodejs.org/job/node-test-pull-request/3349/

insightfuls commented:
Is there anything holding this back from being merged? Keen to see it get into the codebase if possible.

insightfuls commented:
Not wanting to nag, but is there any chance of merging this? Is there anything you'd like me to do? It would be nice if this could get into a release.

jasnell commented:
@insightfuls ... looks like it fell through the cracks. The PR needs to be rebased and updated and likely would need another round of review. /cc @Fishrock123 @MylesBorins

insightfuls commented:
OK, I've rebased it onto current master and resolved the extremely minor conflict that arose (code was equivalent but differently expressed).

insightfuls commented:
Does something more need to happen here to get this merged and released? :-)

gibfahn commented:
cc/ @Fishrock123, is this okay to land?

insightfuls commented:
Any news? Any way we can move this forward?

addaleax commented:
@insightfuls The linter rules have changed since this PR was last updated, we don’t allow blank lines at the ends of files anymore, which is why the CI failed (the ARM failure is only a display error). It would be cool if you could update this PR with that; otherwise the person who merges this PR can do that for you.

insightfuls commented:
Done, @addaleax. I've also rebased onto current master and resolved the conflict that arose in test/README.md. Hopefully it will land this time around!

insightfuls commented:
> why are there no doc changes for the new method?

As per the [description](https://github.com/nodejs/node/pull/7360#issue-161601412) and your [previous comment](https://github.com/nodejs/node/pull/7360#pullrequestreview-16723565), there are no doc changes because this fixes the implementation to conform to the existing documentation.

> shouldn't this be labelled semver-minor?

I haven't the foggiest. I would have expected this to be the kind of answer you would give me, not vice versa. I'm happy to help, but I can't really arbitrate these kinds of things. If you/the community want me to do something, just let me know explicitly what to do and/or point me to some docs, please, and I'll get to it. Sorry if this is something I should know. It's been so long since I read the information about contributing, etc., that it's pretty hazy.

sam-github commented:
@Fishrock123 you assigned this to yourself, does that mean you want to review? I think your assignment blocks Evan's wonderful Chrome plugin from working. If you don't want to review, unassign, and I will land.

sam-github commented:
@insightfuls sorry, when a month goes by, my memory fails me.

I'm going to call it a bug, and label it as such, since having an `unref()` without a `ref()` makes no sense at all, and it was doced to exist, as you say.

insightfuls commented:
In case it helps to decide on semver-related things, note that currently if you call `.ref()` on a TTY stream you get an exception thrown by internal code, i.e. some code internal to the library tries to call something that doesn't exist and throws, it doesn't produce an exception intended for consumption.

insightfuls commented:
No worries, @sam-github. That's one reason I'm keen to get this landed rather than lose momentum and circle the airport again. :-)

thefourtheye commented on tools/test.py:
```diff
@@ -697,6 +697,7 @@ def Execute(args, context, timeout=None, env={}, faketty=False):
     fd_in = fd_err = fd_out
     pty_out = out_master
   else:
+    fd_in = None
```
`fd_in` is set to zero two lines below this. Is this really necessary here?

addaleax replied:
Right – I think the `fd_in = 0` line should be removed, though?

thefourtheye replied:
or we can leave this file unchanged. `fd_in` will be falsy either way.

addaleax replied:
@thefourtheye Hmm, sorry. I thought `fd_in = None` would actually close the fd, but it doesn’t appear to do that… so yes, this line can probably be removed.

insightfuls replied:
I guess someone made a similar fix to me and I didn't notice when I rebased. I'm not very accomplished with Git. Fixed.

thefourtheye commented on test/pseudo-tty/ref_keeps_node_running.js:
```diff
@@ -0,0 +1,28 @@
+require('../common');
+
+const TTY = process.binding('tty_wrap').TTY;
+const isTTY = process.binding('tty_wrap').isTTY;
```
You can simply do

```js
const { TTY, isTTY } = process.binding('tty_wrap');

insightfuls replied:
Done.

Fishrock123 commented:
The linter failed in that CI though, was that fixed?

insightfuls commented:
I did fix the linting issue picked up by CI. I ran `make test` and it seemed to run the linter at the end of the test run and reported no problems. I'll check again with `make lint`. I'll also look into the other comments.

addaleax commented:
New CI: https://ci.nodejs.org/job/node-test-commit/8052/

insightfuls commented:
Branch updated and pushed. Double-checked linting and tests are all OK, and performed the small amount of code tidy-up requested in recent comments.

insightfuls commented:
I've added a commit that adjusts all the `refed` tests to call strictEqual directly. See what you think.

thefourtheye commented on test/parallel/test-handle-wrap-isrefed.js:
```diff
@@ -3,99 +3,128 @@
+  strictEqual(Object.getPrototypeOf(timer._handle).hasOwnProperty('hasRef'),
+              true, 'timer_wrap: hasRef() missing');
+  strictEqual(timer._handle.hasRef(),
+              false, 'timer_wrap: unref() ineffective');
```
Nit: There was no `unref` before this.

insightfuls replied:
Yes there was; it's just a line further up.

Apparently timers don't have `_handle` defined prior to `unref()` being called, so we cannot check `hasRef()` exists, nor assert an initially `refed` state for them. Instead we check the existence of `hasRef()` and the `unrefed` state after already calling `unref()`.

thefourtheye replied:
Oh cool. Thanks for clarifying :-)

thefourtheye commented:
Why `test/pseudo-tty/ref_keeps_node_running.out` and `test/pseudo-tty/test-handle-wrap-isrefed-tty.out` are necessary?

addaleax commented:
@thefourtheye The test runner expects them, `test/pseudo-tty` is like `test/message` and asserts against the script output.

thefourtheye commented:
Oh okay. Thanks @addaleax :-)

insightfuls commented:
(Just fixed up the author of the commit.)

addaleax commented:
> (Just fixed up the author of the commit.)

@insightfuls Are you sure you didn’t mean the committer of the commit? It might be splitting hairs but I’d like to be sure which email you actually wanted since you explicitly brought it up (`…@users.noreply.github.com` or `dev@…`?).

insightfuls commented:
It doesn't really matter, but the github one is what I intended, and I figured it is probably more sensible if it's at least consistent for the related commits.

addaleax commented:
Landed in 3c92200d8b5b40c86e69b07d87551c55070c96a5 :tada: Thanks for the PR!

insightfuls commented:
Thanks a lot!

Fishrock123 commented:
In hindsight, this shouldn't have this been either:
a. in a minor?
b. not mentioned under "notable changes"?

MylesBorins commented:
Should this be backported?
@Fishrock123 you mention that this should might have been better as a minor... do you think it should land that was into LTS (if it lands)

MylesBorins commented:
ping

gibfahn commented:
Should this be backported to v6.x-staging? If yes please follow [the guide](https://github.com/nodejs/node/blob/master/doc/guides/backporting-to-release-lines.md) and raise a backport PR, if no let me know or add the `dont-land-on` label.

MylesBorins commented:
ping re: backport
